### PR TITLE
refactor dma_memcpy_nd to use `OffsetSizeAndStrideOpInterface`

### PR DESCRIPTION
**Pull request by makslevental**
## TL;DR

This PR refactors `ipu.dma_memcpy_nd` to use [`OffsetSizeAndStrideOpInterface`](https://github.com/llvm/llvm-project/blob/1949fe90bfccaa13b4ecbce33de3123824b9a150/mlir/include/mlir/Interfaces/ViewLikeInterface.td#L33). This enables supporting both operands and attributes for `offsets`/`lengths`/`strides`. This means you can write IR like

```mlir
func.func @sequence(%arg0: memref<2304xi32>, %arg1: memref<2304xi32>, %arg2: memref<2304xi32>) {
  aiex.ipu.dma_memcpy_nd(0, 0, %arg2[0, 0, 0, 0][1, 1, 36, 64][0, 0, 64]) {id = 0 : i64, metadata = @outOF_L2L3} : memref<2304xi32>
  aiex.ipu.dma_memcpy_nd(0, 0, %arg0[0, 0, 0, 0][1, 1, 36, 64][0, 0, 64]) {id = 1 : i64, metadata = @inOF_L3L2} : memref<2304xi32>
  aiex.ipu.sync {channel = 0 : i32, column = 0 : i32, column_num = 1 : i32, direction = 0 : i32, row = 0 : i32, row_num = 1 : i32}
  return
}
```

or

```mlir
func.func @bobsyouruncle(%arg0: memref<64xi32>, %arg1: memref<32xi32>, %arg2: memref<64xi32>) {
  %c0_i64 = arith.constant 0 : i64
  %c1_i64 = arith.constant 1 : i64
  %c64_i64 = arith.constant 64 : i64
  aiex.ipu.dma_memcpy_nd(0, 0, %arg0[%c0_i64, %c0_i64, %c0_i64, %c0_i64][%c1_i64, %c1_i64, %c1_i64, %c64_i64][%c0_i64, %c0_i64, %c0_i64]) {id = 0 : i64, metadata = @objFifo_in0} : memref<64xi32>
  aiex.ipu.dma_memcpy_nd(0, 0, %arg2[%c0_i64, %c0_i64, %c0_i64, %c0_i64][%c1_i64, %c1_i64, %c1_i64, %c64_i64][%c0_i64, %c0_i64, %c0_i64]) {id = 1 : i64, metadata = @objFifo_out0} : memref<64xi32>
  aiex.ipu.sync {channel = 0 : i32, column = 0 : i32, column_num = 1 : i32, direction = 0 : i32, row = 0 : i32, row_num = 1 : i32}
  return
}
```

and both will get munged to the same thing: an array of integers.

## Why

This simplifies the current cpp, which looks like this

```cpp
    SmallVector<uint32_t, 4> offsets(4, 0);
    SmallVector<uint32_t, 4> lengths(4, 1);
    SmallVector<uint32_t, 3> strides(3, 0);

    if (auto c = op.getOffset0().getDefiningOp<arith::ConstantIntOp>())
      offsets[0] = static_cast<uint32_t>(c.value());
    if (auto c = op.getOffset1().getDefiningOp<arith::ConstantIntOp>())
      offsets[1] = static_cast<uint32_t>(c.value());
    if (auto c = op.getOffset2().getDefiningOp<arith::ConstantIntOp>())
      offsets[2] = static_cast<uint32_t>(c.value());
    if (auto c = op.getOffset3().getDefiningOp<arith::ConstantIntOp>())
      offsets[3] = static_cast<uint32_t>(c.value());
    if (auto c = op.getLength0().getDefiningOp<arith::ConstantIntOp>())
      lengths[0] = static_cast<uint32_t>(c.value());
    if (auto c = op.getLength1().getDefiningOp<arith::ConstantIntOp>())
      lengths[1] = static_cast<uint32_t>(c.value());
    if (auto c = op.getLength2().getDefiningOp<arith::ConstantIntOp>())
      lengths[2] = static_cast<uint32_t>(c.value());
    if (auto c = op.getLength3().getDefiningOp<arith::ConstantIntOp>())
      lengths[3] = static_cast<uint32_t>(c.value());
    if (auto c = op.getStride1().getDefiningOp<arith::ConstantIntOp>())
      strides[0] = static_cast<uint32_t>(c.value());
    if (auto c = op.getStride2().getDefiningOp<arith::ConstantIntOp>())
      strides[1] = static_cast<uint32_t>(c.value());
    if (auto c = op.getStride3().getDefiningOp<arith::ConstantIntOp>())
      strides[2] = static_cast<uint32_t>(c.value());
```

to this

```cpp
    llvm::SmallVector<int64_t, 3> strides = llvm::map_to_vector(
        llvm::reverse(op.getMixedStrides()),
        [](OpFoldResult s) { return getConstantIntValue(s).value(); });
    llvm::SmallVector<int64_t, 4> lengths = llvm::map_to_vector(
        llvm::reverse(op.getMixedSizes()),
        [](OpFoldResult s) { return getConstantIntValue(s).value(); });
    llvm::SmallVector<int64_t, 4> offsets = llvm::map_to_vector(
        llvm::reverse(op.getMixedOffsets()),
        [](OpFoldResult s) { return getConstantIntValue(s).value(); });
```

and also the python bindings from this

```python
class IpuDmaMemcpyNd(IpuDmaMemcpyNdOp):
    """Specialize IpuDmaMemcpyNdOp class constructor to take python integers"""

    def __init__(self, metadata, bd_id, mem, offsets=None, lengths=None, strides=None):
        if strides is None:
            strides = [0] * 3
        if lengths is None:
            lengths = [0] * 4
        if offsets is None:
            offsets = [0] * 4
        symMetadata = FlatSymbolRefAttr.get(metadata)
        iTy = IntegerType.get_signless(32)
        x = 0
        y = 0
        intX = arith.ConstantOp(iTy, IntegerAttr.get(iTy, x))
        intY = arith.ConstantOp(iTy, IntegerAttr.get(iTy, y))
        valueOffsets = []
        valueLengths = []
        valueStrides = []
        for i in offsets:
            valueOffsets.append(arith.ConstantOp(iTy, IntegerAttr.get(iTy, i)))
        for i in lengths:
            valueLengths.append(arith.ConstantOp(iTy, IntegerAttr.get(iTy, i)))
        for i in strides:
            valueStrides.append(arith.ConstantOp(iTy, IntegerAttr.get(iTy, i)))
        super().__init__(
            metadata=symMetadata,
            id=bd_id,
            x=intX,
            y=intY,
            memref=mem,
            offset3=valueOffsets[0],
            offset2=valueOffsets[1],
            offset1=valueOffsets[2],
            offset0=valueOffsets[3],
            length3=valueLengths[0],
            length2=valueLengths[1],
            length1=valueLengths[2],
            length0=valueLengths[3],
            stride3=valueStrides[0],
            stride2=valueStrides[1],
            stride1=valueStrides[2],
        )
```

to this

```python
class IpuDmaMemcpyNd(IpuDmaMemcpyNdOp):
    """Specialize IpuDmaMemcpyNdOp class constructor to take python integers"""

    def __init__(
        self,
        metadata,
        bd_id,
        mem,
        offsets: MixedValues = None,
        lengths: MixedValues = None,
        strides: MixedValues = None,
    ):
        x = 0
        y = 0
        if offsets is None:
            offsets = [0] * 4
        if lengths is None:
            lengths = [0] * 4
        if strides is None:
            strides = [0] * 3
        dynamic_offsets, _packed_offsets, static_offsets = _dispatch_mixed_values(
            offsets
        )
        dynamic_lengths, _packed_lengths, static_lengths = _dispatch_mixed_values(
            lengths
        )
        dynamic_strides, _packed_strides, static_strides = _dispatch_mixed_values(
            strides
        )
        super().__init__(
            x,
            y,
            mem,
            dynamic_offsets,
            dynamic_lengths,
            dynamic_strides,
            static_offsets,
            static_lengths,
            static_strides,
            metadata,
            bd_id,
        )
```

## How

This would've been a simple matter of implementing `OffsetSizeAndStrideOpInterface` i.e.

```cpp
    ::mlir::OperandRange getSizes();
    ::llvm::ArrayRef<int64_t> getStaticSizes();
    static unsigned getOffsetSizeAndStrideStartOperandIndex();
    static std::array<unsigned, 3> getArrayAttrMaxRanks();
```

if it weren't for the fact that `OffsetSizeAndStrideOpInterface` [assumes/verifies/enforces](https://github.com/llvm/llvm-project/blob/1949fe90bfccaa13b4ecbce33de3123824b9a150/mlir/lib/Interfaces/ViewLikeInterface.cpp#L52-L58) `strides.rank == lengths.rank`. 

The workaround is some pretty "advanced" usage of interfaces that boils down to

1. Subclassing both `OffsetSizeAndStrideOpInterface` and `OffsetSizeAndStrideOpInterface::Trait` in C++ in order to override `OffsetSizeAndStrideOpInterface::Trait::verifyTrait` to skip the `strides.rank == lengths.rank` check. 
    - The reason you have to do this is because interface inheritance in tablegen just means appending traits, not subclassing, so you can't override trait methods in "subinterfaces".
2. Using [`OpInterfaceTrait`](https://github.com/llvm/llvm-project/blob/edae8f6ce29a980d83761f59f81b88167a0fd815/mlir/include/mlir/IR/Interfaces.td#L36) to expose the pure C++ interface to tablegen.
3. Using `i64` instead of `i32` indices (the interface requires either `index` or `i64` and `i64` seems more straightforward).

## N.B.

- Currently only `arith.constant` operands are supported (and this is verified). In the future, if we want to support fully dynamic operands, this will be straightforward.
- There is a stacked [PR](https://github.com/Xilinx/mlir-aie/pull/905) that chronologically predates the current version of this PR where I rename `lengths` -> `wraps` and `steps` -> `strides` on `dma_memcpy_nd` and `writebd_shimtile`. In light of this current version I propose we actually rename `lengths` -> `sizes` to completely align with `OffsetSizeAndStrideOpInterface`'s naming conventions.
- This PR also updates `objectFifo` -> `objectfifo` in `reference_designs/ipu-xrt/vision_pipelines` which wasn't being tested because OpenCV is not installed on the minisforum runner (i've tested them on my minisforum).

### CODE REVIEW

**Review by stephenneuendorffer — COMMENTED** [review]
This generally makes sense to me, but I'm curious what people who are actively using this think.

**Comment by makslevental** [comment]
> This generally makes sense to me, but I'm curious what people who are actively using this think.

Current users are basically completely unaffected - the existing python bindings were already turning python constants into `arith.constant`. Now the op just accepts those same python constants and turns them into attrs instead of `arith.constant` ops.

**Comment by makslevental** [comment]
The yays have it.

**Comment by AndraBisca** [comment]
This is really nice! Thank you for all of the detailed explanations as well @makslevental, they are very helpful!
